Learning section for NORTH (#198)
FAIRmat-NFDI/nomad-docs: docs/explanation/north.md

as Docker containers and registered with NOMAD. Several such tools, maintained by FAIRmat,
are already available as NOMAD plugins. These tools are based on either Jupyter- or desktop-based images.
Custom containers can also be created and connected to the service.

<!--## How to connect and use specific NORTH tools in a NOMAD deployment
TODO move this part as part of a how-to
Relevant for a given NOMAD deployment is always the specific configuration YAML file that defines which NORTH tools are included. The starting point and example for such a configuration file is the respective [default.yaml file](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/blob/develop/nomad/config/defaults.yaml) within the NOMAD main repository.-->

<!--## Defining official & community tools
TODO what the term official tool means should be documented, does official mean just something that is in any of the projects tracked by the automated plugin collection algorithm, e.g., FAIRmat-NFDI or nomad-coe, or does official mean more, i.e., demands for sophisticated documentation and maintenance strategy by NOMAD/FAIRmat officials-->

## Official base images

> Using base image within the [`nomad-north-desktop-base`](https://github.com/FAIRmat-NFDI/nomad-north-desktop-base) repository is the recommended best practice to build a NORTH tool if that demands a graphical user interface. Its desktop-base image builds on the [Jupyter Docker Stacks project](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html), offering

Reviewer: ```suggestion
Using the base image within the [`nomad-north-desktop-base`](https://github.com/FAIRmat-NFDI/nomad-north-desktop-base) repository is the recommended best practice to build a NORTH tool if that demands a graphical user interface. Its desktop-base image builds on the [Jupyter Docker Stacks project](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html), offering
```

## Introduction

NORTH (NOMAD Remote Tools Hub) is a NOMAD service of data parsing and analysis tools

> which runs in isolated containerized environments that connect to NOMAD's data storage.

Reviewer: In hindsight, "data parsing" (which I also originally used) may not be the best use case to highlight since we have the whole NOMAD parsing setup.

I also find it somewhat weird to say "a service of tools which runs". It is not the overall service that is run in container, but rather the NORTH tools themselves.

What do you think of
```suggestion
NORTH (NOMAD Remote Tools Hub) is a NOMAD service for running tools in isolated containerized environments that connect to NOMAD's data storage.
```

These reproducible and secure functionalities are accessible via the web browser.
NORTH provides a standardized way to run heterogeneous tools that are written in different
programming languages and with different dependencies, without coupling them directly
to the NOMAD core services.

## User perspective

From a user’s perspective, NORTH can be used for running complex or tool-specific analyses
directly on data that is stored inside NOMAD, be this API-retrievable input from NOMAD entries

> or data from ones uploads. Results achieved within the container can be written back

Reviewer: Can be other people's uploads as well. So I would suggest
```suggestion
directly on data that is stored inside NOMAD, be this API-retrievable input from NOMAD entries
or data from NOMAD uploads. Results achieved within the container can be written back
```

(using the [NOMAD API](../howto/manage/program/api.md)) as derived data, metadata, or artifacts.
Running a reprocessing of an upload afterwards allows for an indexing of these results obtained
with NORTH, provided that the software tools in NORTH write using data structures and schemas

> that NOMAD understands.

Reviewer: One could even trigger the processing through the API..

The connection between the container and the NOMAD file system removes the need
for copying and downloading large datasets. Container images remove the need for installing
analysis software locally and assure consistent execution independent of the user's operating
system. Instead, the tools are executed remotely within the infrastructure
that a NOMAD deployment provides. These capabilities of NORTH are especially important in
collaborative research settings enabling other users to rerun the same analysis, or apply
existent parameterizations to new data. Provided a server installation and cloud configuration
is used, this also enables to spawn multiple instances of the same tool.

> The central deployment of NOMAD is one example that substantiates these capabilities.

Reviewer: I think this last sentence does not provide any substance for the reader, it's fairly obvious from the docs that the central deployment is the one that users would check out first.

Unlike the central NOMAD services (which are optimized for data ingestion, storage, indexing,
and search), NORTH is designed for computationally intensive, tool-specific, or rapidly evolving
analyses. These include custom analysis software, post-processing steps, domain-specific
workflows, or tools with complex or conflicting dependencies that cannot reasonably be embedded
into the NOMAD core.

## Architecture

Architecturally, NORTH acts as a dedicated execution layer that is separate from the
NOMAD core services. Tools are executed as Docker containers, with each run isolated
from other tools and the NOMAD services themselves. NOMAD is responsible for launching
NORTH and configuring the tools, while NORTH is responsible for managing container startup,
execution, and teardown, all managed through its JupyterHub service.

**TODO clarify which and then give the link to the repo behind JupyterHub service, one of these?**

> https://gitlab.mpcdf.mpg.de/nomad-lab/north/jupyterhub or https://github.com/FAIRmat-NFDI/nomad-north

Reviewer: https://github.com/FAIRmat-NFDI/nomad-north is the correct one. I am not sure if we want to expose this in the docs at all, this is fairly internal.

## Using existing tools

NORTH does not implement analysis logic itself. Instead, it executes tools that are packaged
as Docker containers and registered with NOMAD. A number of such tools, maintained by FAIRmat,
are already provided as plugins to the NOMAD ecosystem. These tools are based either on jupyter-
or on desktop-based images. Creating ones own containers and connecting these to the service

> is also possible.

Reviewer: Sounds a bit clunky, how about

```suggestion
as Docker containers and registered with NOMAD. Several such tools, maintained by FAIRmat, are already available as plugins as NOMAD plugins. These tools are based on either Jupyter- or desktop-based images. Custom containers can also be created and connected to the service.
```

<!--## How to connect and use specific NORTH tools in a NOMAD deployment

> TODO move this part as part of a how-to

Reviewer: Add this to #200 instead.

Relevant for a given NOMAD deployment is always the specific configuration YAML file that defines which NORTH tools are included. The starting point and example for such a configuration file is the respective [default.yaml file](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/blob/develop/nomad/config/defaults.yaml) within the NOMAD main repository.-->

<!--## Defining official & community tools
TODO what the term official tool means should be documented, does official mean just something that is in any of the projects tracked by the automated plugin collection algorithm, e.g., FAIRmat-NFDI or nomad-coe, or does official mean more, i.e., demands for sophisticated documentation and maintenance strategy by NOMAD/FAIRmat officials-->

> ## Official base images

Reviewer: I think it's good to mention them here, we should then point to this section in #200

Using base image within the [`nomad-north-desktop-base`](https://github.com/FAIRmat-NFDI/nomad-north-desktop-base) repository is the recommended best practice to build a NORTH tool if that demands a graphical user interface. Its desktop-base image builds on the [Jupyter Docker Stacks project](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html), offering
a [conda](https://anaconda.org) environment with a jupyter notebook and graphical user interface via a light-weight [xfce](https://www.xfce.org/) desktop environment. Ubuntu Linux is the base layer of this docker image, connecting the container to the operation system of the host computer.

<!--REMOVE WHEN REFACTORING COMPLETED
previously called [base-desktop](https://gitlab.mpcdf.mpg.de/nomad-lab/north/base-desktop/-/blob/main/Dockerfile?ref_type=heads) -->

> Using the base image within the [cookiecutter-nomad-plugin](https://github.com/FAIRmat-NFDI/cookiecutter-nomad-plugin) repository is the recommended best practice to build a NORTH tool from if it does not demand a graphical user interface. Note that this repository is the template for building

Reviewer: I think we should rather point to https://github.com/FAIRmat-NFDI/nomad-plugin-template, that's the one most plugin developers are familiar with and that we point them to frequently. The cookiecutter one is just the "backend" for this.

Note also that they should still use it if they want to use a graphical user interface (this needs to be explained in #200). In that case, they need to make changes to the Dockerfile, but the folder structure/entry point etc. should be from https://github.com/FAIRmat-NFDI/nomad-plugin-template and https://github.com/FAIRmat-NFDI/cookiecutter-nomad-plugin.

NOMAD plugins.

<!--REMOVE WHEN REFACTORING COMPLETED
<!--[jupyter notebook] is the one as of 2026/01/19 shipped with nomad-FAIR. -->
<!--gitlab-registry.mpcdf.mpg.de/nomad-lab/nomad-distro/jupyter:develop -->
<!--https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-distro/container_registry/1462?orderBy=NAME&sort=asc&search[]=latest -->

## Specific NORTH tools

Consult the reference section of the documentation to find details about individual NOMAD plugins and the eventual NORTH tools these provide.

> <!--MOVE THESE HERE INDIVIDUAL DOCUMENTATION TO A NORTH ENTRY POINT SPECIFIC SECTION IN THE PLUGINS

Reviewer: For me it's fine to have all these commented-out docs, we just have to be very strict about removing them asap.

[the GitLab repositories of current NORTH tools](https://gitlab.mpcdf.mpg.de/nomad-lab/north)

## Code location & refactoring history

[Since its addition to NOMAD as a service](https://joss.theoj.org/papers/10.21105/joss.05388), the backend and docker images behind NORTH tools have been in a process of significant refactoring. Not only did the docker base images evolve substantially but also most services of NOMAD were refactored into plugins. NORTH tools with demands for a graphical user interface that were initially based on [Webtop](https://docs.linuxserver.io/images/docker-webtop/) are becoming replaced by an image that is based on the [Jupyter Docker Stacks project](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html).

> In this process, the storage location of the code that FAIRmat developed for NORTH changed. The [initial](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-remote-tools-hub) location had already been considered deprecated. **But this is not necessarily mentioned in all repos** The content in the location that was used [subsequently](https://gitlab.mpcdf.mpg.de/nomad-lab/north) is also in a process of refactoring whereby individual NORTH tools get anchored as a service to an **official NOMAD plugin (we need to distinguish these clean in the documentation, expectation management, and whose responsibility is it to maintain these?** that matches closest thematically and functionally.

Reviewer: I feel like this is too much information about the history. I would focus on the status as-is or rather as it soon will be (e.g. north tools based on `nomad-north-desktop-base`, configured as entry points).

Author: Reduced this, moved it downwards, marked it for deprecation ones the refactoring is complete, status as is now addressed in official base images.

## Defining official & community tools

> **A definition ??? Official and community NORTH tools, the distinction should be documented; each official NORTH tool need to have a documentation and maintenance strategy, i.e., a responsible person from FAIRmat/NOMAD. We should add a section what community NORTH tool maintainers should do if they want to have their container listed here.**

Reviewer: We are doing all this dicussion in #186. I would say we do not need to do it here. Rather, if this is not clear from #186, we should put this sentence there. What do you think @JFRudzinski? Is it clear enough what we mean by "official plugin/tool"?

Author: No, a specific summary of the development history of NORTH is not part of #186, any other part of the nomad documentation, or the scientific literature, shortened it though, see comment above.

The constant refactoring that NORTH has experienced brought changes to locations of docker images and updates of the images. Relevant for a given NOMAD deployment is always the specific configuration YAML file that defines which NORTH tools are included. The starting point and example for such a configuration file is the respective [default.yaml file](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/blob/develop/nomad/config/defaults.yaml) within the NOMAD main repository.

### nomad-north-desktop-base

[`nomad-north-desktop-base`](https://github.com/FAIRmat-NFDI/nomad-north-desktop-base), previously called [base-desktop](https://gitlab.mpcdf.mpg.de/nomad-lab/north/base-desktop/-/blob/main/Dockerfile?ref_type=heads), is the [Jupyter Docker Stacks project](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html) base docker image that we consider as the best practice to build NORTH tools from if these need support for GUI elements.

### jupyter-notebook
<!--### nomad-north-jupyter-base-->
[jupyter notebook] is the one as of 2026/01/19 shipped with nomad-FAIR. TODO
<!-- gitlab-registry.mpcdf.mpg.de/nomad-lab/nomad-distro/jupyter:develop -->

> <!-- https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-distro/container_registry/1462?orderBy=NAME&sort=asc&search[]=latest -->

Reviewer: This is incorrect. As we decided now with Area D, we will _not_ have a "base jupyter tool" from which to build your own tool. Rather, the one that comes from the cookiecutter for nomad plugins should be used.

Author: changed that, see official base images

> ## Official tools (need functional check)

Reviewer: This is all super valuable information that should be kept somewhere. The question is where.

Lauri suggested to not put these explanations about the "official" tools here. The problem with these manual explanations is that somebody needs to maintain this page forever. As @JFRudzinski poitned out, we already have an automated approach for documenting existing plugins. As NORTH tools are just defined through entry points from now on, they will show up there as well.

So basically, these descriptions should go to the entry points of the individual tools. Any further explanation should go to the documentation of the plugin itself.

Reviewer: We can just link to the page in #186 and to the [`NOMAD Plugins App`](https://nomad-lab.eu/prod/v1/oasis/gui/search/plugins) and tell users to filter for NORTH tools there.

Author: Agree to proceed with this suggested, added text to hook in the page from 186. [NOMAD plugin registry](../examples/plugin_registry.md)

<!-- shall we not add pynxtools to the jupyter base image?-->

### jupyter

**The [`jupyter`](https://gitlab.mpcdf.mpg.de/nomad-lab/north/jupyter) container should be replaced by the non-GUI jupyter image from Rubel renamed to `nomad-north-jupyter` ?.**

### jupyterhub

**The [`jupyterhub`](https://gitlab.mpcdf.mpg.de/nomad-lab/north/jupyterhub) container should be replaced by [nomad-north](https://github.com/FAIRmat-NFDI/nomad-north) ?**

### spm and sts
[`spm`](https://gitlab.mpcdf.mpg.de/nomad-lab/north/spm) and [`sts`](https://gitlab.mpcdf.mpg.de/nomad-lab/north/sts) and are two jupyter-based NORTH tools offering examples for scanning probe microscopy and scanning tunneling probe spectroscopy. **The containers  are to be renamed `nomad-north-spm` and `nomad-north-sts` respectively, currently both are functionally similar, ideal would be to add a functionality that goes beyond mere conversion, to warrant relevance, and difference, for sure the containers should be moved to the [pynxtools-spm](https://fairmat-nfdi.github.io/pynxtools-spm/) plugin.**

### voila

> [`voila`](https://gitlab.mpcdf.mpg.de/nomad-lab/north/voila) is a jupyter-based NORTH tool simplying the usage of jupyter notebooks via [voila](https://github.com/voila-dashboards/voila). The tool is currently developed by the Helmholtz Zentrum Berlin (HZB) and can thus be considered a user-provided NORTH tool. **voila as a tool can be helpful, but that container needs refactoring, firstly, renaming, at all an official one, if so why does the image have HZB-specific customizations, who will maintain this.**

Reviewer: As discussed, we can probably provide this as a basic container stripped of the HZB-specific customizations and if they want to maintain their own version (ideally using this generic one as base) they should do it in their own plugin.

Author: jup, I commented out all content and explanation for individual north tools as I agree these should go to the individual plugin's documentation, e.g., there in a north entry point section

Guidance on developing and registering custom NORTH tools is available in the how-tos:
[How to develop NORTH tools](../howto/plugins/types/north_tools.md).

## Data schema interoperability

Beyond its importance for managing the expectations as to how NOMAD reacts to and what
individual NORTH tools offer, it is important to know that using different data schemas
in a NORTH tool compared to the NOMAD deployment does not always come without challenges.

Users may face issues when reprocessing uploads when results have been stored in these
uploads from a NORTH tool analysis whose application wrote a different version of the
data schema than used for the NOMAD deployment. Eventual incompatibilities in parts or
entirely are possible. Having now individual NORTH tools as optional functionality
additions of a NOMAD plugin offload the responsibility to reduce the incompatibilities

> between different schema versions to the plugin developers.

Reviewer: A bit clunky again, how about
```suggestion

Users may encounter issues when reprocessing uploads that contain results from a NORTH tool analysis, if the tool wrote data using a schema version different from the one used in the NOMAD deployment. This can lead to partial or complete incompatibilities. By providing individual NORTH tools as optional extensions of a NOMAD plugin, the responsibility for minimizing schema incompatibilities is shifted to the plugin developers.
```

> ## NORTH refactoring history

Reviewer: I think we could put this in an [info or warning admonition](https://github.com/FAIRmat-NFDI/nomad-docs/blob/develop/docs/writing_guide.md?plain=1#L82) for now. That way we still keep this information, but don't overload any new users reading this.
